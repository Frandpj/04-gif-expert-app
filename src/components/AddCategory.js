import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState(''); // Si se deja vacío es undefined y marca un warning, se puede poner un string vacío
    
    /* El input se queda con un valor asignado
    y esta función hace que se pueda cambiar ese valor */
    const handleInputChange = (e) => {
        setInputValue(e.target.value)
    }

    // Hace que el formulario funcione sin actualizar la página e imprime un mensaje
    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length > 2) {
            setCategories(cast => [...cast, inputValue]);
            setInputValue('');
        }
    }

    return (
        <form onSubmit={ handleSubmit }>
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )
}

// Hace que la función setCategories sea obligatoria para que el componente funcione
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}
