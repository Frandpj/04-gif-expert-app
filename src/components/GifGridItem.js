import React from 'react'

export const GifGridItem = ({ id, title, url }) => {

    console.log(id, title, url);
    return (
        <div>
            <img src={ url } alt={ title } />
            <p>{ title }</p>
        </div>
    )
}
