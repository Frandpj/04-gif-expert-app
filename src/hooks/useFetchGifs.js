import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {

        getGifs(category)
            .then(imgs => {
                setTimeout(() => {
                    setState({
                        data: imgs,
                        loading: false
                    });
                }, 3000);
        });

    }, [category]);

    // A los 3 segundos se cargan los datos y deja de cargar (loading: false)
    // setTimeout(() => {
    //     setState({
    //         data: [1, 2, 3, 4, 5, 6, 7],
    //         loading: false
    //     })
    // }, 3000);

    return state; // { data: [] , loading: true};

}
