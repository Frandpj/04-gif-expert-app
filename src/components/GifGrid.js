import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
// import { GifGridItem } from './GifGridItem';
// import { getGifs } from '../helpers/getGifs';

export const GifGrid = ({ category }) => {

    // const [images, setImages] = useState([])
    const { loading } = useFetchGifs();

    console.log(loading);

    /* useEffect con un segundo argumento array vacío hace que este código solo 
    se ejecute una vez, si en el array añadimos el parámetro, si por algún casual
    cambia la categoría (parámetro) volverá a disparar la función */
    // useEffect(() => {
    //     getGifs(category)
    //         .then(setImages)
    // }, [ category ])

    return (
        <>
            <h3>{ category }</h3>

            { loading  ? 'Cargando...' : 'Data cargada'}
            {/* <div className="card-grid">
                Tarea, imprimir lista con key = id y el titulo = title de la función getGifs 
                {
                    images.map(img => (
                        <GifGridItem 
                            key={ img.id }
                            // Se pasa el objeto tal cual para que en el otro lado se pueda desestructurar
                            // Lo que llega al otro lado ya no es variable.img si no img nada más
                            { ...img }
                        />
                    ))
                }
            </div> */}
        </>
    )
}
