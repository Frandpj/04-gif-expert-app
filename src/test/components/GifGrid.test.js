import React from 'react'
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en <GifGrid/>', () => {

    const category = 'Haikyuu';

    test('debe de mostrar el componente correctamente', () => {

        // Cuando se llame useFetchGifs dentro del componente retorna el valor
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category} />);
        
        expect(wrapper).toMatchSnapshot();

    })

    test('debe de mostrar items cuando se cargan imÃ¡genes useFetchGifs', () => {

        const  gifs = [{
                id: 'ABC',
                url: 'https://localhost/cualquiera/cosa.jpg',
                title: 'Cualquier cosa'
            },
            {
                id: '123',
                url: 'https://localhost/cualquiera/cosa.jpg',
                title: 'Cualquier cosa'
            }
        ];

        // Se manda los valores con mock
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category} />);

        // expect(wrapper).toMatchSnapshot();

        expect(wrapper.find('p').exists()).toBe(false);

        // Selecciona el elemento del componente
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length)
        
    })
    
    
    
})
