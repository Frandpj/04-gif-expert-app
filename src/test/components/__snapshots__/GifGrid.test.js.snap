// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid/> debe de mostrar el componente correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Haikyuu
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
